osu.Game.Rulesets.Osu/OsuDifficulty/Preprocessing/OsuDifficultyBeatmap.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) 2007-2017 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using System.Collections;
+using System.Collections.Generic;
+using osu.Game.Rulesets.Osu.Objects;
+
+namespace osu.Game.Rulesets.Osu.OsuDifficulty.Preprocessing
+{
+    public class OsuDifficultyBeatmap : IEnumerable<OsuDifficultyHitObject>
+    {
+        private readonly IEnumerator<OsuDifficultyHitObject> difficultyObjects;
+        private readonly Queue<OsuDifficultyHitObject> onScreen = new Queue<OsuDifficultyHitObject>();
+
+        public OsuDifficultyBeatmap(List<OsuHitObject> objects)
+        {
+            // Sort HitObjects by StartTime - they are not correctly ordered in some cases.
+            // This should probably happen before the objects reach the difficulty calculator.
+            objects.Sort((a, b) => a.StartTime.CompareTo(b.StartTime));
+            difficultyObjects = createDifficultyObjectEnumerator(objects);
+        }
+
+        public IEnumerator<OsuDifficultyHitObject> GetEnumerator()
+        {
+            do
+            {
+                // Add upcoming notes to the queue until we have at least one note that had been hit and can be dequeued.
+                // This means there is always at least one note in the queue unless we reached the end of the map.
+                bool hasNext;

```

Comment:
It might be nicer to move `hasNext` into the loop and to just explicitly break if `hasNext` is false. As it stands now, the loop will always be existed at the end once `hasNext` becomes false anyways.